synapse/tests/test_lib_ast.py

```diff
@@ -651,6 +696,248 @@ class AstTest(s_test.SynTest):
             self.stormIsInPrint('arg2: world', msgs)
             self.stormIsInPrint('arg3: goodbye', msgs)
 
+            # Basic function chaining
+            q = '''
+            function inner() {
+                $lib.print("inner vertex")
+                return ("foobarbazbiz")
+            }
+
+            function outer() {
+                return ($inner())
+            }
+
+            $output = $outer()
+            $lib.print($output)
+            '''
+            msgs = await core.streamstorm(q).list()
+            self.stormIsInPrint('inner vertex', msgs)
+            self.stormIsInPrint('foobarbazbiz', msgs)
+
+            # return a directly called function
+            q = '''
+            function woot(arg1) {
+                return ( $($arg1 + 1) )
+            }
+
+            function squee(arg2) {
+                return ($woot($arg2))
+            }
+            $output = $squee(17)
+            $lib.print('output is {a}', a=$output)
+            '''
+
+            msgs = await core.streamstorm(q).list()
+            self.stormIsInPrint('output is 18', msgs)
+
+            # recursive functions
+            q = '''
+            function recurse(cond, count) {
+                if $( $cond = 15 ) {
+                    return ($count)
+                }
+                return ($recurse( $($cond - 1), $($count + 1) ))
+            }
+            $output = $recurse(21, 0)
+            $lib.print('final recursive output is {out}', out=$output)
+            '''
+
+            msgs = await core.streamstorm(q).list()
+            self.stormIsInPrint('final recursive output is 6', msgs)
+
+            # return a function (not a value, but a ref to the function itself)
+            q = '''
+            function toreturn() {
+                $lib.print("toreturn called")
+                return ("foobar")
+            }
+
+            function wrapper() {
+                return ($toreturn)
+            }
+
+            $func = $wrapper()
+            $lib.print("this should be first")
+            $output = $func()
+            $lib.print("got {out}", out=$output)
+            '''
+            msgs = await core.streamstorm(q).list()
+            expected = ['this should be first', 'toreturn called', 'got foobar']
```

Comment:
why not just call stormIsInPrint 3 times?

Reply:
fixed. it was supposed to testing the order of the messages, and it was, but i made it more robust.